src/GraphQL/Types/Composite/InputObjectGraphType.cs

```diff
@@ -60,20 +61,79 @@ public virtual object ParseDictionary(IDictionary<string, object?> value)
         }
 
         /// <inheritdoc/>
-        public virtual bool IsValidDefault(object value) => value is TSourceType;
+        public virtual bool IsValidDefault(object value)
+        {
+            if (value is not TSourceType)
+                return false;
+
+            foreach (var field in Fields)
+            {
+                if (!field.ResolvedType!.IsValidDefault(GetFieldValue(field, value)))
+                    return false;
+            }
+
+            return true;
+        }
 
         /// <summary>
         /// Converts a value to an AST representation. This is necessary for introspection queries
-        /// to return the default value for fields of this input object type. This method may throw an exception
-        /// or return <see langword="null"/> for a failed conversion.
+        /// to return the default value for fields of this input object type. Also AST representation
+        /// is used while printing schema as SDL.
+        /// <br/>
+        /// This method may throw an exception or return <see langword="null"/> for a failed conversion.
+        /// <br/><br/>
+        /// The default implementation returns <see cref="GraphQLNullValue"/> if <paramref name="value"/>
+        /// is <see langword="null"/> and <see cref="GraphQLObjectValue"/> filled with the values
+        /// for all input fields except ones returning <see cref="GraphQLNullValue"/>.
         /// <br/><br/>
-        /// The default implementation always throws an exception. It is recommended that this method be
-        /// overridden to support introspection of fields of this type that have default values. This method
-        /// is not otherwise needed to be implemented.
+        /// Note that you may need to override this method if you have already overrided <see cref="ParseDictionary"/>.
         /// </summary>
-        public virtual GraphQLObjectValue? ToAST(object value)
+        public virtual GraphQLValue ToAST(object? value)
```

Comment:
Such a default implementation is similar to the one that is now resides in `SchemaPrinter` where string value is concatenated manually - https://github.com/graphql-dotnet/graphql-dotnet/blob/master/src/GraphQL/Utilities/SchemaPrinter.cs#L382